packages/library-sync/src/commonMain/kotlin/io/realm/kotlin/mongodb/exceptions/SyncExceptions.kt

```diff
@@ -44,22 +51,73 @@ public open class SyncException : AppException {
  *
  * @see io.realm.kotlin.mongodb.sync.SyncConfiguration.Builder.errorHandler
  */
-public class UnrecoverableSyncException : SyncException {
-    internal constructor(message: String) : super(message)
-}
+public class UnrecoverableSyncException internal constructor(message: String) :
+    SyncException(message)
 
 /**
  * Thrown when the type of sync used by the server does not match the one used by the client, i.e.
  * the server and client disagrees whether to use Partition-based or Flexible Sync.
  */
-public class WrongSyncTypeException : SyncException {
-    internal constructor(message: String) : super(message)
-}
+public class WrongSyncTypeException internal constructor(message: String) : SyncException(message)
 
 /**
  * Thrown when the server does not support one or more of the queries defined in the
  * [io.realm.kotlin.mongodb.sync.SubscriptionSet].
  */
-public class BadFlexibleSyncQueryException : SyncException {
-    internal constructor(message: String) : super(message)
+public class BadFlexibleSyncQueryException internal constructor(message: String) :
+    SyncException(message)
+
+/**
+ * Thrown when the server undoes one or more client writes. Details on undone writes can be found in
+ * [writes].
+ */
+public class CompensatingWriteException internal constructor(
+    message: String,
+    compensatingWrites: Array<CoreCompensatingWriteInfo>
+) : SyncException(message) {
+    /**
+     * List with all the reversed writes that have triggered this exception.
+     */
+    public val writes: List<CompensatingWriteInfo> = compensatingWrites.map {
+        CompensatingWriteInfo(
+            reason = it.reason,
+            objectName = it.objectName,
+            primaryKey = it.primaryKey.asRealmAny(),
+        )
+    }
+
+    private fun RealmValue.asRealmAny(): RealmAny? = when (getType()) {
```

Comment:
It is a slight different piece of code, it is supposed to be only valid for PK and there is no way to actually convert to RealmObjects, we lack of access to the mediator and so.

Reply:
But you could drop all this into a `primitiveRealmValueToAny` (or something like that) and rework the `fromRealmValue` of `realmAnyConverter` to call this new method in all other cases that for `RLM_TYPE_LINK` then we wouldn't have duplicate implementations and wouldn't miss to keep `RealmValue`->`RealmAny` conversion consistent if we update it. If that is to cumbersome for this PR (seems like there is also diverging behavior when you cannot convert it), then at least put this conversion method in `Converters.kt` with a descriptive name, so that we wouldn't redo it if we need something similar later on.

Reply:
Looks way better now 🏆